File: python/4.py
#!/usr/bin/env python
LARGEST = 999
SMALLEST = 100

def is_palindrome(i):
    return str(i) == str(i)[::-1]

def has_three_digit_factors(i):
    for j in range(LARGEST, SMALLEST - 1, -1):
        if i % j == 0:
            quotient = i / j
            if SMALLEST <= quotient <= LARGEST:
                return True
    return False

def main():
    for i in range(LARGEST**2, SMALLEST, -1):
        if is_palindrome(i):
            if has_three_digit_factors(i):
                print(i)
                break

if __name__ == '__main__':
    main()

File: rust/src/bin/4.rs
fn main() {
    let interim: i32 = 999;
    let largest: i32 = interim.pow(2);
    let smallest: i32 = 100;
    for i in (smallest..largest).rev() {
        if is_palindrome(i) {
            if has_three_digit_factors(i) {
                    println!("{}", i);
                    break;
            }
        }
    }
}

fn is_palindrome(i: i32) -> bool {
    let str_i = i.to_string();
    let str_j: String = str_i.chars().rev().collect();
    return str_i == str_j;
}

fn has_three_digit_factors(i: i32) -> bool {
    for j in (500..999).rev() {
        if i % j == 0 {
            let quotient = i / j;
            if (quotient > 100) && (quotient <= 999) {
                return true;
            }
        }
    }
    return false;
}
